### PR TITLE
[agg_v2] Add more documentation and unbounded constructor

### PR DESCRIPTION
**Pull request by igor-aptos**
(did the same already on aggregator fallback PR as well)

### CODE REVIEW

**Review by vusirikala — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/aggregator_v2/aggregator_v2.move [hunk 1]
```diff
@@ -1,4 +1,14 @@
-/// This module provides an interface for aggregators (version 2).
+/// This module provides an interface for aggregators (version 2). Aggregators are
+/// similar to unsigned integers and support addition and subtraction (aborting on
+/// underflow or on overflowing a custom upper limit). The difference from integers
+/// is that aggregators allow to perform both additions and subtractions in parallel
+/// across multiple transactions, enabling parallel execution. For example, if the
+/// first transaction is doing `try_add(X, 1)` for aggregator `X`, and the second is
+/// doing `try_sub(X,3)`, they can be executed in parallel avoiding a read-modify-write
+/// dependency.
+/// However, reading the aggregator value (i.e. calling `read(X)`) is an expensive
```
Comment (gelash) [inline aptos-move/framework/aptos-framework/sources/aggregator_v2/aggregator_v2.move:9]: is an expensive operation that also reduces parallelism, and should be avoided as much as possible?

File: aptos-move/framework/aptos-framework/sources/aggregator_v2/aggregator_v2.move [hunk 2]
```diff
@@ -63,19 +90,28 @@ module aptos_framework::aggregator_v2 {
     }
 
     /// Returns a value stored in this aggregator.
+    /// Note: This operation prevents parallelism of the transaction that calls it.
```
Comment (gelash) [inline aptos-move/framework/aptos-framework/sources/aggregator_v2/aggregator_v2.move:93]: let's maybe expand, it prevents parallelism with other transactions that update the aggregator? (maybe also that it is expensive, as expensive and parallelism conflict both are complementary true)

File: aptos-move/framework/aptos-framework/sources/aggregator_v2/aggregator_v2.move [hunk 3]
```diff
@@ -63,19 +90,28 @@ module aptos_framework::aggregator_v2 {
...
     public native fun read<IntElement>(aggregator: &Aggregator<IntElement>): IntElement;
 
+    /// Returns a wrapper of a current value of an aggregator
+    /// Unlike read(), this enables parallel execution of transactions.
```
Comment (gelash) [inline aptos-move/framework/aptos-framework/sources/aggregator_v2/aggregator_v2.move:97]: it is fast and maybe "avoids sequential dependencies w. / "

File: aptos-move/framework/aptos-framework/sources/aggregator_v2/aggregator_v2.move [hunk 4]
```diff
@@ -63,19 +90,28 @@ module aptos_framework::aggregator_v2 {
...
 
+    /// Creates a snapshot of a given value.
+    /// Useful for when object is sometimes created via snapshot() or string_concat(), and sometimes directly.
     public native fun create_snapshot<Element: copy + drop>(value: Element): AggregatorSnapshot<Element>;
 
+    /// NOT YET IMPLEMENTED, always raises EAGGREGATOR_FUNCTION_NOT_YET_SUPPORTED.
     public native fun copy_snapshot<Element: copy + drop>(snapshot: &AggregatorSnapshot<Element>): AggregatorSnapshot<Element>;
 
+    /// Returns a value stored in this snapshot.
+    /// Note: This operation prevents parallelism of the transaction that calls it.
```
Comment (gelash) [inline aptos-move/framework/aptos-framework/sources/aggregator_v2/aggregator_v2.move:108]: ditto, let's slightly expand on "prevents parallelism"

File: aptos-move/framework/aptos-framework/sources/aggregator_v2/aggregator_v2.move [hunk 5]
```diff
@@ -85,9 +121,10 @@ module aptos_framework::aggregator_v2 {
     //     features::change_feature_flags(fx, vector[feature], vector[]);
 
     //     let snapshot = create_snapshot(42);
-    //     let snapshot2 = copy_snapshot(&snapshot);
+    //     // copy not yet supported
+    //     // let snapshot2 = copy_snapshot(&snapshot);
     //     assert!(read_snapshot(&snapshot) == 42, 0);
-    //     assert!(read_snapshot(&snapshot2) == 42, 0);
+    //     // assert!(read_snapshot(&snapshot2) == 42, 0);
```
Comment (gelash) [inline aptos-move/framework/aptos-framework/sources/aggregator_v2/aggregator_v2.move:127]: double commented out?

File: aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs [hunk 6]
```diff
@@ -229,6 +229,52 @@ fn native_create_aggregator(
...
+/***************************************************************************************************
+ * native fun create_unbounded_aggregator<IntElement: copy + drop>(): Aggregator<IntElement>;
+ **************************************************************************************************/
+
+ fn native_create_unbounded_aggregator(
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    if !context.aggregator_api_enabled() {
```
Comment (gelash) [inline aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs:242]: this block (checking context for enabled) should be re-used in many methods, can we make it a method/macro call?

File: aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs [hunk 7]
```diff
@@ -229,6 +229,52 @@ fn native_create_aggregator(
...
+            Type::U64 => u64::MAX as u128,
+            _ => {
+                return Err(SafeNativeError::Abort {
+                    abort_code: EUNSUPPORTED_AGGREGATOR_TYPE,
+                })
+            },
+        }
+    };
+
+    let value_field_value = if context.aggregator_execution_enabled() {
```
Comment (gelash) [inline aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs:263]: can we re-use the later part with native_create_aggregator?

File: aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs [hunk 8]
```diff
@@ -229,6 +229,52 @@ fn native_create_aggregator(
     ]))])
 }
 
+
+/***************************************************************************************************
+ * native fun create_unbounded_aggregator<IntElement: copy + drop>(): Aggregator<IntElement>;
+ **************************************************************************************************/
+
+ fn native_create_unbounded_aggregator(
```
Comment (gelash) [inline aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs:237]: not super important here, but for my curiosity, do we have ways to unit test these native interfaces (with context)?

**Review by gelash — APPROVED** [review]

File: aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs (hunk 8 above)
Comment (georgemitenkov) [reply aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs:237]: I think Move unit tests should cover all scenarios?

Comment (georgemitenkov) [reply aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs:237]: @igor-aptos can we have a test with supported/unsupported types?

**Comment by igor-aptos** [comment]
merged into https://github.com/aptos-labs/aptos-core/pull/10397

**Pull request closed without merging**